### ISSUE TITLE
sysread returned an error: An existing connection was forcibly closed by the remote host

### ISSUE THREAD

**Issue by yaroslav-shabalin** [glasswalk3r/siebel-monitoring-tools#5]
Hello Alceu!
With introduction of v0.20 there is an error during connection establishing process on MS Windows. I've attached the log file. As far as I understand there is some delay between initial banner and srvrmgr prompt. After processing the banner daemon tries to sysread STDOUT one more time and fails with an error "An existing connection was forcibly closed by the remote host". On the same machine and with the same connection parameters v0.19 works well.
[log.txt](https://github.com/glasswalk3r/siebel-monitoring-tools/files/40056/log.txt)

**Comment by glasswalk3r** [comment]
Hi,
Most probably because I "turned on" again ```_check_error``` invocation, which I disable by mistake in the past.
Windows does have a problem by doing IPC with srvrmgr (probably with any other program anyway). I tried to use a non-document trick by using socketpair to simulate a socket in place of regular file handle over the program.
For some reason, if the socketpair is not working for some time, it will return this error. If you set the ```$ENV{SIEBEL_SRVRMGR_DEVEL}``` variable and run the test *t/DaemonHeavy.t* you will get the same error... and since there is a mock of srvrmgr used by these tests... there is no much reason for a timeout.
I don't have a fix for this yet, and to be honest, I'm not looking for one either since **IPC in Windows is simple unreliable**... but there are possible workarounds and I'm quite open to apply pull requests for it.
One workaround is disable ```_check_error``` and see if happens... but I'm not betting on this one. I've already left a script running with a instance of Siebel::Srvrmgr::Daemon::Heavy for a whole week on Linux and didn't got this error... and executing thru it was happening only from 15 to 15 minutes... on Windows this would die much sooner.
Another option is setup a retry on Siebel::Srvrmgr::Daemon::Heavy... and create another child process. But for this case of yours, I don't think it will work either.
Finally... replace Siebel::Srvrmgr::Daemon::Heavy from Siebel::Srvrmgr::Daemon::Light when running on Windows. This could be implemented by creating a instance of Siebel::Srvrmgr::Daemon::Light when running in Windows. I'm sure it will work (no IPC at all), but the trade-off is opening sequentially connections to the Siebel Enterprise.
I'm working in migrating Siebel-Srvrmgr-Exporter to Dist::Zilla and put it available on CPAN as a separated distribution. If you're willing to create a patch for it, please wait till I finish that. Some more fixes are in order too, but it will be easier for a merge after I migrate this to Dist::Zilla (and away from *examples* directory)

**Comment by glasswalk3r** [comment]
I just finished migrating Siebel-Srvrmgr to Dist::Zilla.
From now on it should be easier to start working with this issue. Please let me know your findings (if any!).

**Comment by glasswalk3r** [comment]
Some references on this error:
WSAECONNRESET (10054)

* Translation: Connection reset by peer.
* Description: An existing connection was forcibly closed by the remote host. This error typically occurs if the peer program on the remote host is suddenly stopped, the host is restarted, or the remote host uses a hard close. See setsockopt (Wsapiref_94aa.asp) for more information about the SO_LINGER option on the remote socket. This error may also result if a connection was broken because of keep-alive activity that detects a failure while one or more operations are in progress. Operations that were in progress fail with WSAENETRESET. Subsequent operations fail with WSAECONNRESET

https://support.microsoft.com/en-us/kb/819124

If you run the stress_DaemonHeavy.t test under xt folder, you will get the same error, even though srvrmgr-mock.pl doesn't make any connection with a Siebel server (and thus is quite fast).

Can you please try to set this SO_LINGER option to the socketpair and see what happens?

```perl
my $linger = pack("ii", 1, 0);
$sock->sockopt(SO_LINGER, $linger) or die "sockopt: $!\n";
```

Maybe the log from srvrmgr tells us something more about (although it is clearly a problem with the socketpair, not the program).

**Comment by yaroslav-shabalin** [comment]
Hi,
What I have tried:
1. Commenting `_check_error` didn't help.
2. Using Siebel::Srvrmgr::Daemon::Light instead of Siebel::Srvrmgr::Daemon::Heavy seems to be a good workaround. No errors and runs quite fast.
3. Adding `SO_LINGER` option. Haven't managed to make it working.
Just to be sure, I invoke the `sockopt` just before opening connection, don't I? Something like that:
```perl
sub _mswin_open3 {

    my $cmd_ref = shift;

    my ( $inRead,  $inWrite )  = Siebel::Srvrmgr::IPC::_mswin_pipe();
    my ( $outRead, $outWrite ) = Siebel::Srvrmgr::IPC::_mswin_pipe();
    my ( $errRead, $errWrite ) = Siebel::Srvrmgr::IPC::_mswin_pipe();

    my $linger = pack("ii", 1, 0);
    $outWrite->sockopt(SO_LINGER, $linger) or die "sockopt: $!\n";

    my $pid = open3(
        '>&' . fileno($inRead),
        '<&' . fileno($outWrite),
        '<&' . fileno($errWrite),
        @{$cmd_ref}
    );

    return ( $pid, $inWrite, $outRead, $errRead );
}
```
BTW, hasn't it be `pack("vv", 1, 0)` according to [linger structure reference](https://msdn.microsoft.com/ru-ru/library/windows/desktop/ms739165%28v=vs.85%29.aspx) (unsigned short little-endian)?

Anyway the best result I got is that script exits on the first atempt to call `sysread()` not on the second. There are some errors in attached srvrmgr log-file (that is from export_comps normal run where I at least get the banner). Starting srvrmgr directly at the same time shows no errors and connection is successfully established.
[srvrmgr.txt](https://github.com/glasswalk3r/siebel-monitoring-tools/files/47512/srvrmgr.txt)

**Comment by glasswalk3r** [comment]
Thanks for running all those tests.

Well, ```check_error``` is really important for Siebel::Srvrmgr::Daemon subclasses because it avoids exceptions from the parser, which will lead to errors that are hard to troubleshoot if something goes wrong before getting any valid output. Let's leave it untouched if it's not the problem.

I think you're correct about setting SO_LINGER with ```sockopt```, you should use "vv" as the value for ```pack```. **But** I think you're hitting other problem. By the error message:
```
SBL-SCM-00018: Could not open connection to Siebel Gateway configuration store
```
I understand that you have a authentication problem with srvrmgr. You see, I changed the way Siebel::Srvrmgr::Daemon::Heavy handles passwords because this class is intended to be used in long running process. If I execute srvrmgr with -p option, then it would disclosure the password and be a security issue. So, now this class is omitting the -p options and providing the password directly to STDIN of srvrmgr program. Maybe this is not working on Windows as expected (well, what does it anyway?). I think it's worth checking on that.

Then, maybe it's worth to try SO_LINGER option, but as I said before, socketpair is a hack in this case. **And**, as you already checked, Siebel::Srvrmgr::Daemon::Light is a good replacement for Siebel-Srvrmgr-Exporter on Windows. It will work several connections to get the work done, but it's a good trade-off instead not working at all. Right now I think trying to fix Siebel::Srvrmgr::Daemon::Heavy is simply not worth the time and effort.

If you're willing to provide a patch, please do it. I just ask you to selective choose one subclass of Siebel::Srvrgmr::Daemon depending on the value of $Config{osname}: if it's windows, setup a Siebel::Srvrmgr::Daemon::Light class. That should be enough handle.

Beware that I have updated all the projects (moved them from Extutils::MakeMaker to Dist::Zilla), so you will want to get a new clone/pull changes.

**Comment by glasswalk3r** [comment]
Just added a new version into the repository with a conditional control to replace Siebel::Srvrmgr::Daemon::Heavy by Siebel::Srvrmgr::Daemon::Light. That should avoid completely the issue with IPC on MS Windows systems.
Could you please validate? You can checkout the code available under tag release-0.24

**Comment by glasswalk3r** [comment]
I'm closing this issue under the assumption that the recent change fix this. There is no issues reported for running Siebel::Srvrmgr::Daemon::Light on MS Windows.
